c++/src/kj/async-io.h

```diff
@@ -168,6 +222,41 @@ class AsyncIoStream: public AsyncInputStream, public AsyncOutputStream {
   // isn't wrapping a file descriptor.
 };
 
+class AsyncBufferedIoStreamWrapper: public AsyncBufferedInputStream, public AsyncIoStream {
```

Comment:
This inheritance hierarchy won't work very well, because you're actually inheriting two copies of `AsyncInputStream`. You could theoretically fix that with virtual inheritance, but that can lead to other issues.

Another weird thing about this class is that if we later introduce a buffered output stream, someone might expect that this class will add buffering in both directions.

Perhaps we should simply change `AsyncIoMessageStream` to optionally accept the input and output as separate objects.

Reply:
`AsyncBufferedInputStream` deliberately doesn't inherit from `AsyncInputStream` to solve the issue of two copies.

I agree about the buffered output stream problem though, and I did think about it but couldn't really come up with a great solution -- I considered was bringing over the `AsyncIoStream` implementation that lets you pass in the input/output sides to the constructor from one of the KJ tests, but this has the problem of not being able to implement the methods that are defined on `AsyncIoStream` itself.

There's also the problem of buffering for `AsyncCapabilityStream` since there is no `AsyncCapabilityInputStream`, if we do figure out a way to implement that -- we'd need to something there to have an equivalent of passing separate input/output objects for `AsyncCapabilityMessageStream`.

Reply:
Oof, `AsyncCapabilityStream` is indeed a problem.

I missed that `AsyncBufferedInputStream` doesn't subclass `AsyncInputStream`, but that creates a different problem... you wouldn't be able to use `kj::dynamicDowncastIfAvailable` to discover if a particular `AsyncInputStream` implements `AsyncBufferedInputStream`.

Reply:
I wonder if just sticking with the `MessageStream` approach might be simpler at this point. Obviously I'm biased because I don't want to throw away code, but it's difficult to come up with a nice interface here that works for normal byte streams and capability streams at this level of abstraction... If you don't want to commit to that interface in capnp, perhaps we can keep it as part of the runtime?

Reply:
I guess I should take another look at that version and see if I'm OK with it.